update rocm-recipes with master version

keerthana4708 opened the pull request:
Signed-off-by: msaikeer <Makala.SaiKeerthana@amd.com>

amd-msaikeer-devops committed:
update rocm-recipes with master version

Signed-off-by: msaikeer <Makala.SaiKeerthana@amd.com>

haampie commented on var/spack/repos/builtin/packages/rocm-openmp-extras/package.py:
```diff
@@ -135,6 +158,13 @@ class RocmOpenmpExtras(Package):
+            url='https://github.com/ROCm-Developer-Tools/aomp-extras.git',
+            destination='rocm-openmp-extras',
+            placement='aomp-extras',
+            when='master')
```
```suggestion
            when='@master')
```

haampie commented on var/spack/repos/builtin/packages/rocm-openmp-extras/package.py:
```diff
@@ -144,6 +174,13 @@ class RocmOpenmpExtras(Package):
+            url='https://github.com/llvm/llvm-project.git',
+            destination='rocm-openmp-extras',
+            placement='llvm-project',
+            when='master')
```
```suggestion
            when='@master')
```

haampie commented on var/spack/repos/builtin/packages/rocm-openmp-extras/package.py:
```diff
@@ -117,6 +124,14 @@ class RocmOpenmpExtras(Package):
+            url='https://github.com/RadeonOpenCompute/ROCm-Device-Libs.git',
+            destination='rocm-openmp-extras',
+            placement='rocm-device-libs',
+            when='master')
```
```suggestion
            when='@master')
```

haampie commented on var/spack/repos/builtin/packages/rocm-openmp-extras/package.py:
```diff
@@ -100,6 +101,12 @@ class RocmOpenmpExtras(Package):
         depends_on('hsa-rocr-dev@' + ver, when='@' + ver)
         depends_on('llvm-amdgpu@{0} ~openmp'.format(ver),
                    when='@' + ver)
+    git = "https://github.com/ROCm-Developer-Tools/aomp.git"
```
move this up

keerthana4708 replied:
ACK

haampie commented on var/spack/repos/builtin/packages/rccl/package.py:
```diff
@@ -64,6 +65,6 @@ def cmake_args(self):
             args.append(self.define('__skip_rocmclang', 'ON'))
 
-        if self.spec.satisfies('@4.5.0:'):
+        if self.spec.satisfies('@4.5.0:') :
```
```suggestion
        if self.spec.satisfies('@4.5.0:'):
```

haampie commented on var/spack/repos/builtin/packages/llvm-amdgpu/package.py:
```diff
@@ -111,7 +110,7 @@ def cmake_args(self):
         ]
         args = []
-        if self.spec.satisfies('@4.3.0:4.5.0'):
+        if self.spec.satisfies('@4.3.0:4.5.0') or self.spec.satisfies('@master'):
```
```suggestion
        if self.spec.satisfies('@4.3.0:4.5.0,master'):
```
or just
```suggestion
        if self.spec.satisfies('@4.3.0:'):
```

haampie replied:
@srekolam any reason there's an upperbound here?

amd-msaikeer-devops committed:
update rocm-recipes with master version

Signed-off-by: msaikeer <Makala.SaiKeerthana@amd.com>

keerthana4708 committed:
Update var/spack/repos/builtin/packages/rocm-openmp-extras/package.py

Co-authored-by: Harmen Stoppels <harmenstoppels@gmail.com>

keerthana4708 committed:
Update var/spack/repos/builtin/packages/rocm-openmp-extras/package.py

Co-authored-by: Harmen Stoppels <harmenstoppels@gmail.com>

keerthana4708 committed:
Update var/spack/repos/builtin/packages/rocm-openmp-extras/package.py

Co-authored-by: Harmen Stoppels <harmenstoppels@gmail.com>

keerthana4708 committed:
Update var/spack/repos/builtin/packages/rccl/package.py

Co-authored-by: Harmen Stoppels <harmenstoppels@gmail.com>

keerthana4708 committed:
Update var/spack/repos/builtin/packages/llvm-amdgpu/package.py

Co-authored-by: Harmen Stoppels <harmenstoppels@gmail.com>

amd-msaikeer-devops committed:
update rocm-recipes with master version

Signed-off-by: msaikeer <Makala.SaiKeerthana@amd.com>

srekolam commented on var/spack/repos/builtin/packages/llvm-amdgpu/package.py:
```diff
@@ -111,7 +110,7 @@ def cmake_args(self):
             'compiler-rt'
         ]
         args = []
-        if self.spec.satisfies('@4.3.0:4.5.0'):
```
@haampie , i think we do not the upper bound here... i added for 4.5.0 ..but this applicable for 4.5.2..which i forgot to update. i think if self.spec.satisfies('@4.3.0:'):  will do

keerthana4708 replied:
Changed to  self.spec.satisfies('@4.3.0:'):

amd-msaikeer-devops committed:
update rocm-recipes with master version

Signed-off-by: msaikeer <Makala.SaiKeerthana@amd.com>

amd-msaikeer-devops committed:
update rocm-recipes with master version

Signed-off-by: msaikeer <Makala.SaiKeerthana@amd.com>

haampie commented:
@keerthana4708 please fix the style issues before notifying me

amd-msaikeer-devops committed:
update rocm-recipes with master version

Signed-off-by: msaikeer <Makala.SaiKeerthana@amd.com>

keerthana4708 commented:
> @keerthana4708 please fix the style issues before notifying me

Hi @haampie, Done with style issues

keerthana4708 commented:
> > @keerthana4708 please fix the style issues before notifying me
> 
> Hi @haampie, Done with style issues

Hi @haampie , did you get a chance to review this PR

amd-msaikeer-devops committed:
update rocm-recipes with master version

Signed-off-by: msaikeer <Makala.SaiKeerthana@amd.com>

srekolam reviewed (dismissed):
can you get it reviewed with ethan stewart for rocm-openmp-extras ?

keerthana4708 commented:
> > > @keerthana4708 please fix the style issues before notifying me
> > 
> > 
> > Hi @haampie, Done with style issues
> 
> Hi @haampie , did you get a chance to review this PR

Hi @haampie , Could you please review this PR

srekolam commented:
@haampie  can we get this reviewed please

keerthana4708 commented:
> > > > @keerthana4708 please fix the style issues before notifying me
> > > 
> > > 
> > > Hi @haampie, Done with style issues
> > 
> > 
> > Hi @haampie , did you get a chance to review this PR
> 
> Hi @haampie , Could you please review this PR

Hi @trws , Could you please review this PR

amd-msaikeer-devops committed:
update rocm-recipes with master version

Signed-off-by: msaikeer <Makala.SaiKeerthana@amd.com>

amd-msaikeer-devops committed:
update rocm-recipes with master version

Signed-off-by: msaikeer <Makala.SaiKeerthana@amd.com>

keerthana4708 committed:
Merge branch 'develop' into upstream-develop

keerthana4708 commented:
> can you get it reviewed with ethan stewart for rocm-openmp-extras ?

@estewart08, could you please review rocm-openmp-extras

keerthana4708 commented:
> > > > > @keerthana4708 please fix the style issues before notifying me
> > > > 
> > > > 
> > > > Hi @haampie, Done with style issues
> > > 
> > > 
> > > Hi @haampie , did you get a chance to review this PR
> > 
> > 
> > Hi @haampie , Could you please review this PR
> 
> Hi @trws , Could you please review this PR

Hi  @trws , Could you please review this PR

amd-msaikeer-devops committed:
update rocm-recipes with master version

Signed-off-by: msaikeer <Makala.SaiKeerthana@amd.com>

arjun-raj-kuppala commented:
@haampie can we get this reviewed please

trws commented on var/spack/repos/builtin/packages/hip-rocclr/package.py:
```diff
@@ -25,7 +25,7 @@ def url_for_version(self, version):
         return url.format(version)
 
-    version('master', branch='main')
+    version('master', branch='master')
```
The main branch on here hasn't been touched in 5 years, the master branch in 2, it looks like the only active branch is the default branch, which is develop.  Why add what looks like a dead branch instead of the one that's in use?

keerthana4708 replied:
Hi @trws , have done the necessary changes, Could you please review

trws reviewed (commented):
Have you checked that the branches here are the appropriate branches? I just spot checked and it looks like the branch on rocclr is not for example.

amd-msaikeer-devops committed:
update rocm-recipes with master version

Signed-off-by: msaikeer <Makala.SaiKeerthana@amd.com>

estewart08 commented:
@keerthana4708 It is not part of your changes, but I believe line 72 in llvm-amdgpu/package.py needs to exclude the master version as the fix was brought in by upstream to amd-stg-open and presents a patch already detected issue.

amd-msaikeer-devops committed:
update rocm-recipes with master version

Signed-off-by: msaikeer <Makala.SaiKeerthana@amd.com>

amd-msaikeer-devops committed:
update rocm-recipes with master version

Signed-off-by: msaikeer <Makala.SaiKeerthana@amd.com>

trws reviewed (dismissed):
Alright, I'm good with this now.  @haampie, could you take a look and clear your review if it's ok now?

keerthana4708 commented:
> Alright, I'm good with this now. @haampie, could you take a look and clear your review if it's ok now?

@haampie , could you please review this PR

estewart08 commented on var/spack/repos/builtin/packages/rocm-openmp-extras/package.py:
```diff
@@ -86,8 +86,10 @@ class RocmOpenmpExtras(Package):
+    git = "https://github.com/ROCm-Developer-Tools/aomp.git"
 
     maintainers = ['srekolam', 'arjun-raj-kuppala', 'estewart08']
+    version('master', branch='aomp-dev')
```
branch is incorrect here , please use aomp-epsdb.

keerthana4708 replied:
changes are done

estewart08 commented on var/spack/repos/builtin/packages/rocm-openmp-extras/package.py:
```diff
@@ -113,6 +115,11 @@ class RocmOpenmpExtras(Package):
         depends_on('hsa-rocr-dev@' + ver, when='@' + ver)
         depends_on('llvm-amdgpu@{0} ~openmp'.format(ver),
                    when='@' + ver)
+    if ver == 'master':
```
Indentation seems off here. I believe you can get rid of the if condition and use when=@master instead. Also I think these depends_on clauses should be outside/above the for loop.

keerthana4708 replied:
Done

estewart08 commented on var/spack/repos/builtin/packages/rocm-openmp-extras/package.py:
```diff
@@ -130,6 +137,13 @@ class RocmOpenmpExtras(Package):
             placement='rocm-device-libs',
             when='@' + ver)
 
+        resource(
```
I would prefer the resources for master to be outside of this for loop after the depends_on master changes.

keerthana4708 replied:
Done

estewart08 commented on var/spack/repos/builtin/packages/rocm-openmp-extras/package.py:
```diff
@@ -217,7 +252,8 @@ def patch(self):
         # Openmp adjustments
-        if self.spec.version >= Version('4.5.0'):
+        if (self.spec.version >= Version('4.5.0') or
+            self.spec.version == Version('master')):
```
Correct me if I am wrong, but Spack knows that master > 4.5.0. From my understanding Spack knows that:
```
develop > master > any_numbered_version
```

keerthana4708 replied:
made necessary changes

amd-msaikeer-devops committed:
update rocm-recipes with master version

Signed-off-by: msaikeer <Makala.SaiKeerthana@amd.com>

amd-msaikeer-devops committed:
update rocm-recipes with master version

Signed-off-by: msaikeer <Makala.SaiKeerthana@amd.com>

amd-msaikeer-devops committed:
update rocm-recipes with master version

Signed-off-by: msaikeer <Makala.SaiKeerthana@amd.com>

amd-msaikeer-devops committed:
update rocm-recipes with master version

Signed-off-by: msaikeer <Makala.SaiKeerthana@amd.com>

estewart08 commented on var/spack/repos/builtin/packages/rocm-openmp-extras/package.py:
```diff
@@ -217,11 +258,17 @@ def patch(self):
                 libomptarget.format(src) + '/deviceRTLs/libm/CMakeLists.txt')
 
+        if self.spec.version >= Version('master'):
+           filter_file(
+               '{ROCM_DIR}/amdgcn/bitcode', '{DEVICE_LIBS_DIR}',
+               libomptarget.format(src) + '/libm/CMakeLists.txt')
```
Spacing needs adjusted here as the style check failed.

keerthana4708 replied:
Done with styling part

amd-msaikeer-devops committed:
update rocm-recipes with master version

Signed-off-by: msaikeer <Makala.SaiKeerthana@amd.com>

amd-msaikeer-devops committed:
update rocm-recipes with master version

Signed-off-by: msaikeer <Makala.SaiKeerthana@amd.com>

estewart08 commented:
Thanks for the changes, rocm-openmp-extras looks fine now.

keerthana4708 commented:
@haampie , Could you please review this PR

keerthana4708 commented:
> @haampie , Could you please review this PR

hi @haampie , Could you please review this PR

alalazo commented:
@keerthana4708 Can you rebase the PR to have tests re-run? In general, before pinging maintainers constantly it would be good to have tests passing, thanks!

keerthana4708 committed:
Merge branch 'spack:develop' into upstream-develop

keerthana4708 commented:
> @keerthana4708 Can you rebase the PR to have tests re-run? In general, before pinging maintainers constantly it would be good to have tests passing, thanks!

@alalazo , rebased now again , the review is pending since long time, I have rebased it multiple times.

alalazo commented:
> the review is pending since long time, I have rebased it multiple times.

Thanks, I still see conflicts though :slightly_frowning_face:

keerthana4708 committed:
Merge branch 'develop' into upstream-develop

keerthana4708 commented:
> > the review is pending since long time, I have rebased it multiple times.
> 
> Thanks, I still see conflicts though 🙁

Thanks, fixed the conflict as well

keerthana4708 committed:
Merge branch 'spack:develop' into upstream-develop

keerthana4708 committed:
Merge branch 'spack:develop' into upstream-develop

haampie commented on var/spack/repos/builtin/packages/atmi/package.py:
```diff
@@ -49,9 +50,10 @@ class Atmi(CMakePackage):
+        if not self.spec.satisfies('@master'):
+            return ['-DROCM_VERSION={0}'.format(self.spec.version)]
+        else:
+            return std_cmake_args
```
```suggestion
            return []
```

haampie commented on var/spack/repos/builtin/packages/llvm-amdgpu/package.py:
```diff
@@ -36,17 +35,12 @@ class LlvmAmdgpu(CMakePackage):
-        description='Build LLVM shared library, containing all '
-        'components in a single shared library',
-    )
+    variant('openmp', default=False, description='Enable OpenMP')
```
why change the default?

haampie commented on var/spack/repos/builtin/packages/llvm-amdgpu/package.py:
```diff
@@ -36,17 +35,12 @@ class LlvmAmdgpu(CMakePackage):
+    variant('openmp', default=False, description='Enable OpenMP')
     variant(
-        'link_llvm_dylib',
+        "llvm_dylib",
```
can you undo these changes?

haampie commented on var/spack/repos/builtin/packages/rocm-openmp-extras/package.py:
```diff
@@ -224,11 +266,17 @@ def patch(self):
                 '{ROCM_DIR}/amdgcn/bitcode', '{DEVICE_LIBS_DIR}',
                 libomptarget.format(src) + '/deviceRTLs/libm/CMakeLists.txt')
 
+        if self.spec.version >= Version('master'):
```
I don't think this is going to be a good idea, since `master` is a moving target

haampie commented:
I'm also thinking that `x@master` depends on `y@master` is too strict, I could see you would wanna try a master branch of one of the packages, and leave the rest on the latest stable one? Why not just add a generic `depends_on('y')` without versions. Maybe @cgmb maybe has opinions on whether this makes sense.

keerthana4708 committed:
Merge branch 'spack:develop' into upstream-develop

cgmb commented on var/spack/repos/builtin/packages/rocsolver/package.py:
```diff
@@ -26,6 +26,7 @@ class Rocsolver(CMakePackage):
             size and compile time by adding specialized kernels \
             for small matrix sizes')
 
+    version('master', branch='develop')
```
The two important pre-release branches for rocsolver are `develop` and `master`. Eventually, I'd like to be able to build and install both of them using spack. If we're going to call `develop` 'master' in spack, what will we call `master`?

cgmb reviewed (changes requested):
This is a great feature. Adding this for the math libraries was on my todo list, so it's great to see there was already a PR for it. I'm not entirely clear on how the branches were selected, though.

When I have a moment, I'll elaborate on what the different branches mean for the ROCm math and communication libraries, and what pre-release versions can be expected to work together. @haampie is correct that we can loosen the version restrictions somewhat.

cgmb commented:
On the ROCm math and communication libraries (rocBLAS, hipBLAS, rocFFT, etc.), you will find a number of standard branches:

- **develop**: This is the branch where new features are added. Right now, this is the branch where changes targeting ROCm 5.2 are going.
- **staging**: After some testing, changes are promoted from develop to staging. Changes in this branch are for the same release as develop (whatever that may be at the time).
- **master**: This is the stable or 'mainline' branch. Right now, this is where changes for ROCm 5.1 are going. It will remain that way until a release branch for 5.1 is created, then staging will be promoted to master.

(Other parts of the ROCm stack use different branching schemes. I'm not as familiar with them, unfortunately. I think the process is similar, but the branches may have different names and not all of them are public.)

As a developer working on the math libraries, the two branches I would want to be able to install through spack are `develop` and `master`. For the sake of ensuring that advanced users and outside contributors can build and test these libraries prior to release, we try to ensure that both `develop` and `master` successfully build and run using the toolchain and runtime provided by the most recent public ROCm release. The only caveat is that some new features or extra amdgpu_targets added to the develop or master branches might be disabled when building the pre-release libraries with the already-released toolchain.

So, `develop` and `master` for the ROCm math and communication libraries could reasonably be defined as depending on the most recent public release of the compiler/runtime toolchain (or newer). They will often build and run correctly with older versions, but we don't actually test that, so your mileage may vary. Assuming that spack will default the the newest version, I think @haampie is right that no particular version should be specified as the required version of hip, comgr or rocm-cmake for develop or master.

However, the dependencies between math and communication libraries are stricter. In the packages for rocblas, hipblas, rocsolver, hipsolver, rocfft, hipfft, rocsparse, hipsparse, rocalution, rocprim, hipcub, rocrand, hiprand, rocthrust, and rccl, if any of those libraries depend on each other, then the develop version should depend on develop and the master version should depend on master. That's what this PR is doing for everything, so for that aspect, it's already correct.

keerthana4708 committed:
Merge branch 'spack:develop' into upstream-develop

arjun-raj-kuppala commented:
@haampie 
thoughtbase behind having master version is to be able to setup CI[spack recipes] internally in AMD for complete rocm stack. it will help us in reducing the lead between the actual rocm release and rocm recipe updates for that release.  So, it will be important to have master version of component to depend on its dependencies with master version itself.

cgmb commented on var/spack/repos/builtin/packages/hip/package.py:
```diff
@@ -19,7 +19,7 @@ class Hip(CMakePackage):
 
     maintainers = ['srekolam', 'arjun-raj-kuppala', 'haampie']
-    version('master', branch='master')
+    version('master', branch='develop')
```
Ok. Now I understand a little better why you're choosing these branches. The HIP develop branch is basically the [**staging**](https://github.com/spack/spack/pull/28500#issuecomment-1040694454) concept I described before. (The HIP mainline branch is not publicly available.)

In the math libraries, the develop branch is also basically equivalent to staging. It's nice that we have some consistency about that! And the mathlibs master branch corresponds to mainline. This actually matches with the [spack version comparison](https://spack.readthedocs.io/en/latest/packaging_guide.html#version-comparison); the develop/staging branch is the newest version, and master/mainline is slightly older than develop/staging but newer than the most recent release.

So, I would argue that rather than naming develop/staging as `master` in spack, we should be naming it as `develop`. For tools and libraries that do have a public mainline branch, we can call that one `master` in spack.

cgmb commented:
> thoughtbase behind having master version is to be able to setup CI[spack recipes] internally in AMD for complete rocm stack. it will help us in reducing the lead between the actual rocm release and rocm recipe updates for that release. So, it will be important to have master version of component to depend on its dependencies with master version itself.

It would be convenient for that use case, but it's possible to request specific dependency versions anyway via the spec or your `packages.yaml` / `spack.yaml`. I suppose the question is if optimizing for the simplicity of the CI comes at a cost of usefulness for other purposes. IDK. I should probably pull your branch and try it out.

EDIT: By the way, I'm thrilled to hear that a CI build for the ROCm Spack packages is being worked on. That will be of significant value to both Spack itself and to our libraries. This PR certainly addresses an important need. There's just some details that need ironing out to ensure the use cases of other stakeholders are adequately considered.

arjun-raj-kuppala commented:
> 

Hey Cory, Did you get a chance to try this PR?
Yes, it would help to have CI for rocm spack recipes, i believe this PR will help us in multiple ways. Considering the complexity involved to have this CI with the internal private repositories. This is close to good solution we have come across. If you have any ideas on how to implement the CI considering the internal private development[multiple cross component rocm dependencies which are developed in parallel] would be of great. May be should we have discussion on slack

cgmb commented:
> Hey Cory, Did you get a chance to try this PR?

I have not. I'll take a look in the next few days, though.

cgmb commented:
@haampie, the concretizer seems to believe that all packages in a dependency tree must use the same version of build-time dependencies (such as rocm-cmake). I'm unclear on whether that's a bug in Spack or expected behaviour.

EDIT: If I'm understanding correctly, support for this was added in #28673. That was merged just a few days ago, so I suppose I should rebase and try again.

alalazo commented:
> I'm unclear on whether that's a bug in Spack or expected behaviour.

Having a unique version of a given build dependency in a DAG has been the expected behavior since the first releases of Spack. #28673 introduced the internal data structure to generalize this behavior, but we need an additional PR to allow for having e.g. multiple `cmake` versions in the same DAG.

cgmb commented:
> I'm also thinking that `x@master` depends on `y@master` is too strict, I could see you would wanna try a master branch of one of the packages, and leave the rest on the latest stable one? Why not just add a generic `depends_on('y')` without versions. Maybe @cgmb maybe has opinions on whether this makes sense.

@haampie, basically the problem I encountered when I explored this was that `rocsolver@master ^hip@5.0.2` can't be concretized because `rocsolver@master` legitimately depends on `rocm-cmake@master`, but `hip@5.0.2` depends on `comgr@5.0.2` which depends on `rocm-cmake@5.0.2`. 

rocm-cmake is pretty good at maintaining backwards-compatibility, so we probably could just change all the `depends_on('rocm-cmake@' + ver, type='build', when='@' + ver)` to `depends_on('rocm-cmake@' + ver + ':', type='build', when='@' + ver)` and then `rocm-cmake@master` would satisfy the requirements of both packages.

For better or for worse, we've specified exact version dependencies for all the ROCm dependencies in Spack. The good thing about this approach is that it means that when you run `spack install rocsolver@4.5.0`, the library is built with the exact same dependency versions that were used to build the official DEB and RPM packages for ROCm 4.5.0. The bad thing is that you cannot build `rocsolver@4.5.0 ^hip@4.3.1`, even when there's no good reason for that. (The test suite builds and runs just fine when you compile that version of rocsolver with that version of hip using basic shell commands instead of using Spack.)

cgmb commented:
I've opened #29857 to relax the rocm-cmake version requirements on our packages, which should help prevent the conflicts when mixing release versions with `@master` versions.

srekolam commented:
@keerthana4708  if this still work in progress  can you move it to draft ..or else, please update your branch with the latest and get the review.

cgmb commented:
> @keerthana4708 if this still work in progress can you move it to draft ..or else, please update your branch with the latest and get the review.

@srekolam, this PR is stuck because of disagreements about design. IMO, we should come to an agreement on the design before updating the code. 

I think my proposal in #30273 addresses both my concerns and those of @haampie. Could you, @keerthana4708, and/or @arjun-raj-kuppala review it to ensure that it meets your needs?

srekolam commented:
i think so..we can close this one .
